markups/asciidoc.py

```diff
@@ -0,0 +1,105 @@
+        common.MODULE_HOME_PAGE: 'https://asciidoc3.org/userguide.html',
+        common.SYNTAX_DOCUMENTATION: 'https://asciidoc3.org/userguide.html'
+    }
+
+    file_extensions = ('.adoc', '.asciidoc')
+    default_extension = '.adoc'
+
+    @staticmethod
+    def available():
+        try:
+            importlib.import_module('asciidoc3')
+            importlib.import_module('lxml')
+        except ImportError:
+            return False
+        return True
+
+    def __init__(self, filename=None):
+        AbstractMarkup.__init__(self, filename)
+        from asciidoc3 import asciidoc3api
+        self.converter_class = asciidoc3api.AsciiDoc3API
+
+    def convert(self, text):
+        import asciidoc3
+        from asciidoc3 import asciidoc3api
+        from lxml import etree
+        infile = StringIO()
+        outfile = StringIO()
+        infile.write(text)
+        infile.seek(0)
+        converter = self.converter_class(
+            asciidoc3.__path__[0] + '/asciidoc3.py')
+        try:
+            converter.execute(infile, outfile)
+        except asciidoc3api.AsciiDoc3Error:
+            # Note: asciidoc3 reports/prints these errors for us.
+            # We catch the exception to avoid multiple reports.
+            # sys.stderr.write(f'asciidoc3 {exp}\n')
+            pass
+        # We can get and display messages with the following:
+        # print(f'messages: {converter.messages}')
+        result = outfile.getvalue()
+        result = result.encode()
+        parser = etree.HTMLParser()
+        root_element = etree.fromstring(result, parser)
+        head_element = root_element.xpath('./head')[0]
+        title_element = root_element.xpath('./head/title')[0]
+        style_element = root_element.xpath('./head/style')[0]
+        javascript_elements = root_element.xpath('./head/script')
+        body_element = root_element.xpath('./body')[0]
+        head = ''
+        for child in head_element.getchildren():
+            head += etree.tostring(
+                child,
+                encoding='unicode',
+                method='html',
+            )
+        body = ''
+        for child in body_element.getchildren():
+            body += etree.tostring(
+                child,
+                encoding='unicode',
+                method='html',
+            )
+        title = title_element.text
+        stylesheet = style_element.text
+        if len(javascript_elements) >= 1:
+            javascript = etree.tostring(
+                javascript_elements[0],
+                encoding='unicode',
+                method='html',
+            )
+        else:
+            javascript = ''
+        return ConvertedAsciidoc(head, body, title, stylesheet, javascript)
+
+
+class ConvertedAsciidoc(ConvertedMarkup):
+
+    def __init__(self, head, body, title, stylesheet, javascript):
+        super(ConvertedAsciidoc, self).__init__(
+            body, title, stylesheet, javascript)
+        self.head = head
```

Comment:
Do you need to store `head`? If no, you can eliminate the need for this class and just use ConvertedMarkup.